Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+
+    public record Confirmed(
+        Guid ClientId,
+        DateTimeOffset ConfirmedAt
+    ): ShoppingCartEvent;
+
+    public record Cancelled(
+        DateTimeOffset CanceledAt
+    ): ShoppingCartEvent;
+
+    private ShoppingCartEvent() { }
+}
+
+public abstract record ShoppingCart
+{
+    public record Empty: ShoppingCart;
+
+    public record Pending(ProductItems ProductItems): ShoppingCart;
+
+    public record Closed: ShoppingCart;
+
+    private ShoppingCart() { }
+}
+
+// Primary constructor & TimeProvider
+// Then Collection expressions and advanced pattern matching
+public class ShoppingClassDecider(
+    IProductPriceCalculator priceCalculator,
+    TimeProvider timeProvider
+)
+{
+    public ShoppingCartEvent[] Decide(ShoppingCartCommand command, ShoppingCart state) =>
+        (state, command) switch
+        {
+            (Empty, AddProduct (var productItem, var clientId)) =>
+            [
+                new Opened(clientId, Now),
+                new ProductAdded(priceCalculator.Calculate(productItem).Single(), Now)
+            ],
+
+            (Pending, AddProduct (var productItem, _)) =>
+            [
+                new ProductAdded(priceCalculator.Calculate(productItem).Single(), Now)
+            ],
+
+            (Pending pending, RemoveProduct((var productId, var toRemove, var price) pricedProductItem)) =>
+                pending.ProductItems.TryGetValue($"{productId}{price}", out var quantity) && quantity >= toRemove
+                    ?
+                    [
+                        new ProductRemoved(pricedProductItem, Now)
+                    ]
+                    : throw new InvalidOperationException("Not enough Products!"),
+
+            (Pending, Confirm(var clientId)) =>
+            [
+                new Confirmed(clientId, Now)
+            ],
+
+            (Pending, Cancel) =>
+            [
+                new Cancelled(Now)
+            ],
+
+            (Closed, Confirm) => [],
+
+            (Closed, Cancel) => [],
+
+            _ => throw new InvalidOperationException(nameof(command))
+        };
+
+    private DateTimeOffset Now => timeProvider.GetLocalNow();
+}
+
+public class ShoppingCartModel
+{
+    public required Guid Id { get; init; }
+    public Guid? ClientId { get; set; }
+    public required ShoppingCartStatus Status { get; set; }
+    public List<PricedProductItem> ProductItems { get; init; } = [];
+    public required DateTimeOffset OpenedAt { get; set; }
+    public DateTimeOffset? ConfirmedAt { get; set; }
+    public DateTimeOffset? CanceledAt { get; set; }
+
+    public class PricedProductItem
+    {
+        public required Guid ProductId { get; init; }
+        public required int Quantity { get; set; }
+        public required decimal Price { get; init; }
+        public required DateTimeOffset AddedAt { get; set; }
+        public DateTimeOffset? UpdatedAt { get; set; }
+    }
+
+    public enum ShoppingCartStatus
+    {
+        Pending,
+        Confirmed
```

Comment:
or cancelled?

Reply:
Nope, Cancelled is just removed. It was initially, but I decided to show remove option too.